### ISSUE TITLE
NetworkPolicy with named port is not working when there are more than one port with the same name

### ISSUE THREAD

**Issue by rwilgaard** [cilium/cilium#30003]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

We use named ports in our containers for ingress traffic. When multiple pods have containers with the same port name but with different values, traffic is not allowed.
The issue can be reproduced with the following two pods and a single NetworkPolicy:
```
apiVersion: v1
kind: Pod
metadata:
  name: cilium-np-test01
  namespace: default
  labels:
    app.kubernetes.io/name: cilium-np-test01
    network-rules-egress: cilium-np-test
spec:
  containers:
  - name: debug
    ports:
    - name: http-test
      containerPort: 8000
      protocol: TCP
```
```
apiVersion: v1
kind: Pod
metadata:
  name: cilium-np-test02
  namespace: default
  labels:
    app.kubernetes.io/name: cilium-np-test02
spec:
  containers:
  - name: nginx
    ports:
    - name: http-test
      containerPort: 8080
      protocol: TCP
```
```
---
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  annotations:
  name: allow-egress
  namespace: default
spec:
  podSelector:
    matchLabels:
      network-rules-egress: cilium-np-test
  egress:
  - ports:
    - port: http-test
      protocol: TCP
  policyTypes:
  - Egress
```
The `cilium-np-test01` pod can't reach `cilium-np-test02` when both pods have a `http-test` port with different port values. If I change the `containerPort` on `cilium-np-test01` to 8080 it works or if I remove the `name` from the port on `cilium-np-test01` it works.

### Cilium Version

```
cilium version
Client: 1.14.5 85db28be 2023-12-11T14:30:29+01:00 go version go1.20.12 linux/amd64
Daemon: 1.14.5 85db28be 2023-12-11T14:30:29+01:00 go version go1.20.12 linux/amd64
```

### Kernel Version

```
uname -a
Linux cp03-43a2f386 5.15.0-87-generic #97-Ubuntu SMP Mon Oct 2 21:09:21 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
```

### Kubernetes Version

```
kubectl version
Client Version: v1.28.2
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.27.6+rke2r1
```

### Anything else?

This is a clone of the #24779 issue, that was marked as stale and closed.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by squeed** [comment]
Oof, I wonder if we're resolving the wrong named port. I'll look in to getting this reproduced.

If you're feeling clever, feel free to take a look at the [policy engine implementation.](https://github.com/cilium/cilium/blob/main/pkg/policy/l4.go)

**Comment by squeed** [comment]
Yup, we don't handle this case; we always [collapse to one port](https://github.com/cilium/cilium/blob/main/pkg/endpoint/policy.go#L78).

Right now, the data structure involved in the policy engine doesn't make this easy. Perhaps we can convert these to port ranges once support lands. A bit ugly, but could be worse.

**Comment by Markieta** [comment]
Started noticing this too, particularly on a smaller cluster with fewer nodes where it is more likely pods with the same named port (e.g., `metrics`) land on the same node. A network policy targeting `port: metrics` would translate to BPF policy maps of **only one** of those pods' numeric ports for all mappings.

Any update on this? I assume I should go back to using numeric ports in my network policies in the meantime.

**Comment by andsens** [comment]
Switched from calico recently and had to rewrite some of the policies because of this. It's not a dealbreaker (cilium has global netpols, "entities", and so much more), but man, you could write some really nice netpols with that feature.  
I would love for cilium to find the time and implement this. For example with "metrics" you could then create two networkpolicies that would cover all of your telemetry needs provided that you metrics port is named metrics (the same would go for your ingress controller with e.g. a "web" port):  
```
---
apiVersion: cilium.io/v2
kind: CiliumNetworkPolicy
metadata:
  name: prometheus
spec:
  endpointSelector:
    app.kubernetes.io/name: prometheus
    app.kubernetes.io/component: server
  egress:
  - toEntities:
    - cluster
    toPorts:
    - ports:
      - port: metrics
        protocol: TCP
---
apiVersion: cilium.io/v2
kind: CiliumClusterwideNetworkPolicy
metadata:
  name: metrics-ingress
spec:
  endpointSelector:
    matchLabels: {}
  ingress:
  - fromEndpoints:
    - matchLabels:
        "k8s:io.kubernetes.pod.namespace": prometheus
        app.kubernetes.io/name: prometheus
        app.kubernetes.io/component: server
    toPorts:
    - ports:
      - port: metrics
        protocol: TCP
```

Referencing ports semantically through names just makes everything much more readable, convenient, and ultimately (because of the two former) more secure.

**Comment by WaywardWizard** [comment]
Also an issue i have run into. I looked at the linked GetNamedPort definition linked by squeed. Two assertions/questions:
 - This limitation only applies to spec.egress? (Since named ports are unique to a pod and ingress named ports only need to search ports in pod scope)
 - This applies to 'kind: {Cilium,}NetworkPolicy' (Since they would both use the same policy engine)?

**Comment by rwilgaard** [comment]
Is there any traction on this issue?

**Comment by squeed** [comment]
We've not yet got a good fix for this. I'm hoping that we will have the cycles for v1.18.

**Comment by ErikEngerd** [comment]
Just spent an hour or so troubleshooting a similar issue in a deployment of jupyterhub where the proxy is using a port named 'http' but another pod is also using the same name in the same name. 

Can this be fixed? I can imagine it is also not that easy to fix especially if it must be done efficiently, On the other hand, it should be possible to write a simple utility that is fed with the minimal pod info (such as name, and ports), and then have an API to query the port number based on pod name, port name, and protocol . This can then be tested, also for performance. The current code does this based on name and protocol only (as mentioned above). Integrating this into the cilium code would probably be most of the work. 

It is an important issue. Consider for instance the case of a blue-green deployment where the green pods use the same port name but with a different port number.

**Comment by Doctor-love** [comment]
Bumped into this as well - was pulling my hair out trying to debug some rules. Out of curiosity, what was you approach for finding the root cause, @rwilgaard?

If the issue is tricky to fix, perhaps a warning or similarly could be generated somehow? Can't think of a good way to provide it though.

**Comment by alagoutte** [comment]
> We've not yet got a good fix for this. I'm hoping that we will have the cycles for v1.18.

Always planned to 1.18 ?

**Comment by fristonio** [comment]
@alagoutte this is not going to make it to v1.18. I am hoping to make progress on it for v1.19

**Comment by micke** [comment]
I'm encountering this as well.
But in my case it's not consistent, in some policies it works with named ports in egress rules and in some policies i need to use the actual port number.

For example, this works:

```
apiVersion: cilium.io/v2
kind: CiliumNetworkPolicy
metadata:
  name: prowlarr
spec:
  endpointSelector: {}
  ingress:
    - toPorts:
      - ports:
        - port: http
      fromEndpoints:
        - matchLabels:
            io.kubernetes.pod.namespace: traefik
            app.kubernetes.io/name: traefik
        - matchLabels:
            io.kubernetes.pod.namespace: sonarr
            app.kubernetes.io/component: sonarr
        - matchLabels:
            io.kubernetes.pod.namespace: radarr
            app.kubernetes.io/component: radarr
  egress:
    - toEndpoints:
        - matchLabels:
            io.kubernetes.pod.namespace: kube-system
            k8s-app: kube-dns
      toPorts:
        - ports:
            - port: "53"
              protocol: UDP
          rules:
            dns:
              - matchPattern: "*"
    - toEndpoints:
        - matchLabels:
            io.kubernetes.pod.namespace: sonarr
            app.kubernetes.io/component: sonarr
        - matchLabels:
            io.kubernetes.pod.namespace: radarr
            app.kubernetes.io/component: radarr
      toPorts:
        - ports:
            - port: http
    - toEntities:
        - world
      toPorts:
        - ports:
            - port: "443"
```

It works even if the port named `http` are different actual port numbers in `sonarr` and `radarr`.

While this does not:

```
apiVersion: cilium.io/v2
kind: CiliumNetworkPolicy
metadata:
  name: radarr
spec:
  endpointSelector:
    matchLabels:
      app.kubernetes.io/name: radarr
      app.kubernetes.io/component: radarr
  ingress:
    - toPorts:
      - ports:
        - port: http
      fromEndpoints:
        - matchLabels:
            io.kubernetes.pod.namespace: traefik
            app.kubernetes.io/name: traefik
        - matchLabels:
            io.kubernetes.pod.namespace: unpackerr
            app.kubernetes.io/component: unpackerr
  egress:
    - toEndpoints:
        - matchLabels:
            io.kubernetes.pod.namespace: kube-system
            k8s-app: kube-dns
      toPorts:
        - ports:
            - port: "53"
              protocol: UDP
          rules:
            dns:
              - matchPattern: "*"
    - toEndpoints:
        - matchLabels:
            io.kubernetes.pod.namespace: prowlarr
            app.kubernetes.io/component: prowlarr
        - matchLabels:
            io.kubernetes.pod.namespace: qbittorrent
            app.kubernetes.io/component: qbittorrent
      toPorts:
        - ports:
            - port: http
    - toEntities:
        - world
      toPorts:
        - ports:
          - port: "443"
```

Even if i split up the two endpoints into two different egress rules it does not work:

```
apiVersion: cilium.io/v2
kind: CiliumNetworkPolicy
metadata:
  name: radarr
spec:
  endpointSelector:
    matchLabels:
      app.kubernetes.io/name: radarr
      app.kubernetes.io/component: radarr
  ingress:
    - toPorts:
      - ports:
        - port: http
      fromEndpoints:
        - matchLabels:
            io.kubernetes.pod.namespace: traefik
            app.kubernetes.io/name: traefik
        - matchLabels:
            io.kubernetes.pod.namespace: unpackerr
            app.kubernetes.io/component: unpackerr
  egress:
    - toEndpoints:
        - matchLabels:
            io.kubernetes.pod.namespace: kube-system
            k8s-app: kube-dns
      toPorts:
        - ports:
            - port: "53"
              protocol: UDP
          rules:
            dns:
              - matchPattern: "*"
    - toEndpoints:
        - matchLabels:
            io.kubernetes.pod.namespace: prowlarr
            app.kubernetes.io/component: prowlarr
      toPorts:
        - ports:
            - port: http
    - toEndpoints:
        - matchLabels:
            io.kubernetes.pod.namespace: qbittorrent
            app.kubernetes.io/component: qbittorrent
      toPorts:
        - ports:
            - port: http
    - toEntities:
        - world
      toPorts:
        - ports:
          - port: "443"
```

The only version that actually works is actually specifying the port numbers under egress like this:

```
apiVersion: cilium.io/v2
kind: CiliumNetworkPolicy
metadata:
  name: radarr
spec:
  endpointSelector:
    matchLabels:
      app.kubernetes.io/name: radarr
      app.kubernetes.io/component: radarr
  ingress:
    - toPorts:
      - ports:
        - port: http
      fromEndpoints:
        - matchLabels:
            io.kubernetes.pod.namespace: traefik
            app.kubernetes.io/name: traefik
        - matchLabels:
            io.kubernetes.pod.namespace: unpackerr
            app.kubernetes.io/component: unpackerr
  egress:
    - toEndpoints:
        - matchLabels:
            io.kubernetes.pod.namespace: kube-system
            k8s-app: kube-dns
      toPorts:
        - ports:
            - port: "53"
              protocol: UDP
          rules:
            dns:
              - matchPattern: "*"
    - toEndpoints:
        - matchLabels:
            io.kubernetes.pod.namespace: prowlarr
            app.kubernetes.io/component: prowlarr
      toPorts:
        - ports:
            - port: "9696"
    - toEndpoints:
        - matchLabels:
            io.kubernetes.pod.namespace: qbittorrent
            app.kubernetes.io/component: qbittorrent
      toPorts:
        - ports:
            - port: "8080"
    - toEntities:
        - world
      toPorts:
        - ports:
          - port: "443"
```

I'm not that versed in debugging the policies, but i'm curious if the first policy named `prowlarr` works because it has the same ports under ingress and egress?

**Comment by TheBeeZee** [comment]
The named port lookup is separate (and different) for ingress vs egress. Only the egress named port resolution is actually broken as described in this issue.

In your case there are two "http" ports with conflicting ports (one is 9696, the other is 8080). If you look into Cilium agent logs, you will see "duplicate named port" errors.

This is generated here:
https://github.com/cilium/cilium/blob/main/pkg/types/portmap.go#L22
https://github.com/cilium/cilium/blob/main/pkg/types/portmap.go#L211

The problem is that the policy engine uses a global lookup table to find egress named ports, but expects to find a single egress port when doing the lookup. The solution would be to keep a separate named port map per destination identity, but that is likely a significant refactor, hence the delay.

I might actually look into this, as I'm working on k8s ClusterNetworkPolicy and it breaks the conformance tests.

**Comment by TheBeeZee** [comment]
The fix is now merged into main. Given how long this issue has been open, I think this would be a good candidate to merge into 1.19.

**Comment by andsens** [comment]
@TheBeeZee this is awesome! Thank you for your work on this.  
Being able to add proper semantics to netpols makes them *so* much more readable, really looking forward to using this!

**Comment by joestringer** [comment]
For those following along I opened a discussion thread regarding backporting here: https://github.com/cilium/cilium/pull/44429#issuecomment-4050969782 .
